lib/uklock/include/uk/rwlock.h

```diff
@@ -0,0 +1,304 @@
+#ifndef __UK_RWLOCK_H__
+#define __UK_RWLOCK_H__
+
+#include <uk/config.h>
+
+#include <uk/essentials.h>
+#include <uk/arch/atomic.h>
+#include <stddef.h>
+#include <uk/assert.h>
+#include <uk/wait.h>
+#include <uk/wait_types.h>
+
+#include <uk/print.h>
+
+/* TODO: in future give spin as well as wait options to the user */
+
+#define FLAG_BITS		(5)
+#define READERS_SHIFT		(FLAG_BITS)
+#define FLAG_MASK		((1 << READERS_SHIFT) - 1)
+
+#define UK_RWLOCK_READ		(0x01)
+#define UK_RWLOCK_READ_WAITERS		(0x02)
+#define UK_RWLOCK_WRITE_WAITERS		(0x04)
+#define UK_RWLOCK_WAITERS		(UK_RWLOCK_READ_WAITERS | UK_RWLOCK_WRITE_WAITERS)
+
+#define UK_RW_READERS_LOCK(x)		((x) << READERS_SHIFT | UK_RWLOCK_READ)
+#define UK_RW_ONE_READER		(1 << READERS_SHIFT)
+#define UK_RW_READERS(x)		((x) >> READERS_SHIFT)
+
+#define UK_RW_UNLOCK		(UK_RW_READERS_LOCK(0))
+#define OWNER(x)		(x & ~FLAG_MASK)
+
+#define _rw_can_read(rwlock) \
+	((rwlock & (UK_RWLOCK_READ | UK_RWLOCK_WRITE_WAITERS)) == UK_RWLOCK_READ)
+
+#define _rw_can_write(rwlock) \
+	((rwlock & ~(UK_RWLOCK_WAITERS)) == UK_RW_UNLOCK)
+
+#define _rw_can_upgrade(rwlock) \
+	((rwlock & UK_RWLOCK_READ) && UK_RW_READERS(rwlock) == 1)
+
+#define UK_RWLOCK_INITIALIZER { UK_RW_UNLOCK, 0, 0, 0 }
+
+struct __align(8) uk_rwlock {
+	uintptr_t rwlock;
+#ifdef CONFIG_RWLOCK_WRITE_RECURSE
+	unsigned int write_recurse;
+#endif
```

Comment:
```suggestion
#endif /* CONFIG_RWLOCK_WRITE_RECURSE */
```

Always add a comment to the endif with the condition